storm-core/src/jvm/org/apache/storm/StormTimer.java

```diff
@@ -0,0 +1,274 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm;
+
+import org.apache.storm.utils.Time;
+import org.apache.storm.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Comparator;
+import java.util.Random;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * The timer defined in this file is very similar to java.util.Timer, except
+ * it integrates with Storm's time simulation capabilities. This lets us test
+ * code that does asynchronous work on the timer thread
+ */
+
+public class StormTimer {
+    private static final Logger LOG = LoggerFactory.getLogger(StormTimer.class);
+
+    public interface TimerFunc {
+        public void run(Object o);
+    }
+
+    public static class QueueEntry {
+        public final Long endTimeMs;
+        public final TimerFunc afn;
```

Comment:
The current Timer API allows users to set a "onKill" function to be executed if an exception is thrown in the timer thread.  The "onKill" function take one argument i.e. the throwable that was caught.  So should we use Runnable interface in conjunction with TimerFuncs interface since the "onKill" function cannot be a Runnable as it takes one parameter.  Seems kind of messy if we use the both